### PR TITLE
Speak 'Block moved up/down' after using keyboard actions to move up/down

### LINKED ISSUE #61168: Moving blocks via keyboard: screen readers don't notify operation done
## Short description: 
When using a screen reader, it's not possible to know if blocks have moved up and down, using Gutenberg's keyboard shortcuts ctrl+alt+shift+t -move up-, and ctrl+alt+shift+y -move down- 
## When it works
If a block is moved up and down using arrows, the "move up" and "move down" buttons on blocks toolbar, screen reader correctly announces "block moved from position x to position y".
But to reach the toolbar I'm forced to move the focus away from editing mode, with alt+f10 then navigate to arrows, then move back to editor.
## Steps to reproduce
1. open a blank post or page, then add some blocks
2. focus on the second block. Both navigation and editing mode are affected by the issue. 
3. try to press ctrl+alt+shift+t to move the block upwards one position. Then ctrl+alt+shift+y to move it back in its original place. 
4. Screen reader - tried Jaws, VoiceOver and NVDA- give no feedback about the block's movement. 
## Expected results: 
When pressing ctrl+alt+shift+t or ctrl+alt+shift+y on a block to move it, both in navigation and editing mode, screen reader should announce if the block has moved and where, like it does when I use the arrows. 
## Current results
When pressing ctrl+alt+shift+t or ctrl+alt+shift+y on a block to move it, both in navigation and editing mode, screen reader does not announce if the block has moved and where. It stays silent, and I'm forced to double-check whether or not the blocks have been moved. 
## WordPress and Gutenberg versions
It was present in 17.x branch and persists in latest 18.2.0 build. WordPress 6.5.2 is used.

### PR DESCRIPTION
**Pull request by n2erjo00**
## What?
Describing to screen reader what just happened when user uses keyboard short cuts to move block up or down.

## Why?
Closes #61168

Issue was reported by the linked issue.

## How?

Using `speak` function to speak up after operation is done.

## Testing Instructions
1. Open editor and insert two or three blocks. Blocks don't need be anything fancy
2. Open up screen reader of your choice and select one of the blocks. Move it up with key board shortcut
   - Mac: `⌘ + ⌥ + ⇧ + T or Y`
   - Windows: `CTRL + ALT + SHIFT + T or Y`

### CODE REVIEW

**Review by t-hamano — COMMENTED** [review]
Thanks for the PR!

Simply adding a message may convey incorrect information to the screen reader, because the message will always be read when the shortcut is executed.

For example, the following scenarios need to be considered:

- When a block is at the beginning of the content, it cannot be moved forward
- When a block is at the end of the content, it cannot be moved backward
- Multiple blocks may be selected
- There may be only one block in the content
- In a flex layout, blocks are moved left and right, not up and down

Also, I thinks the message `Block moved {up|down}.` alone does not convey what changes have been made to the content. For example, a message like `Move 2 blocks from position 3 right by one place` is needed.

I think we can use [the `getBlockMoverDescription()` function](https://github.com/WordPress/gutenberg/blob/843e14b0e7768117ca6571bd009d04f522b63374/packages/block-editor/src/components/block-mover/mover-description.js#L37) to cover all of these scenarios.

**Comment by n2erjo00** [comment]
Thanks @t-hamano for the feedback. I'll be taking a look next week

**Comment by t-hamano** [comment]
Thanks for the update!

In my opinion, there is no need to make the processing an external variable. We can probably do it like this:

```javascript
export default function BlockTools( {
	children,
	__unstableContentRef,
	...props
} ) {
	// ...

	function onKeyDown( event ) {
		// ...
		if (
			isMatch( 'core/block-editor/move-up', event ) ||
			isMatch( 'core/block-editor/move-down', event )
		) {
			const clientIds = getSelectedBlockClientIds();
			if ( clientIds.length ) {
				const direction = isMatch( 'core/block-editor/move-up', event )
					? 'up'
					: 'down';
				event.preventDefault();
				const rootClientId = getBlockRootClientId( clientIds[ 0 ] );
				// Here's the logic
				if ( direction === 'up' ) {
					moveBlocksUp( clientIds, rootClientId );
				} else {
					moveBlocksDown( clientIds, rootClientId );
				}

				const description = getBlockMoverDescription(
					// parameters
				);
				speak( description );
			}
		} else if ( isMatch( 'core/block-editor/duplicate', event ) ) {
			// ...
		}
	}
	// ...
}
```

This will also allow us to standardize some of the duplicate processing between `move-up` and `move-down`.

**Comment by n2erjo00** [comment]
@t-hamano Good point much cleaner that way. I implemented the changes

**Comment by t-hamano** [comment]
@n2erjo00 Thanks for the update!

I'm sorry, I missed an important point. `getBlockMoverDescription()` does not tell us the location of the _moved_ block, but where the block is _being moved to_.

Perhaps we need to build a new logic for shortcuts or add a new argument like `isMoved` to `getBlockMoverDescription()`.

Also, I'm trying to check in [this comment](https://github.com/WordPress/gutenberg/issues/61168#issuecomment-2393595213) whether the current block mover reading itself is correct.

Sorry for keeping you waiting, but I'd like to think of an ideal approach.

**Comment by afercia** [comment]
Just noting a previous attempt to add audible messages was tried in https://github.com/WordPress/gutenberg/pull/24894 more than 4 years ago.

I'm not sure these messages should use `getBlockMoverDescription()` by altering the position count. That's meant to be used for the aria-describedby attribute and it would be best to separate the concerns. https://github.com/WordPress/gutenberg/pull/24894  tried to replicate the existing logic and provide meaningful messages for all the cases. 

We could consider to simplify the message but I think two different cases should be considered:

Rationale:
- As an user, when using the buttons I already know what is going to happen. The description provides me a meaningful information, for example:
  - `Move %1$s block from position %2$d down to position %3$d`
  - `Move %1$s block from position %2$d left to position %3$d`
  - `Block %1$s is at the end of the content and can’t be moved down`
- In this case, I don't need to hear again `Block moved to position %2$d down to position %3$d` etc. I only need a confirmation e.g. 'Block moved'. Repeating the new position information in the message would be noisy.
- Instead, when using the keyboard shortcuts, I don't have any information about the initial position. In this case, the confirmation message should provide the new position information.

I do realize distinguishin the two cases is more complex. If it turns out to be too comples, I'd have bo objections with a somehow noisy, redundant message for the buttons.

**Comment by n2erjo00** [comment]
Would the proper course of action to be to create similar function as `getBlockMoverDescription` (maybe `getBlockSpeackDescription` working title) but it would return simplified message: "Block {name} moved {n} steps {direction}" and this PR point of view the return message is then passed to `speak` function

CC @t-hamano @afercia

**Comment by afercia** [comment]
https://github.com/WordPress/gutenberg/pull/24894/files was actually using a dedicated function to provide detailed confirmation messages. It was a little overkill maybe.

@alexstine @talksina from your perspective, when a block is moved would you prefer a detailed confirmation message that announces the new position and direction, which may get a little noisy, or a shorter message? Examples:
- Detailed message: Block Paragraph moved 1 step down.
- Short message: Block moved.

**Comment by alexstine** [comment]
The short message is probably okay considering even "1 step down" really doesn't communicate exactly where the block is.

**Comment by n2erjo00** [comment]
@t-hamano Added short confirmation message after block has been moved up|down

CC @afercia @alexstine

**Comment by t-hamano** [comment]
@afercia @alexstine Thanks for the feedback! So let's use a short message here.

@n2erjo00 I think two changes are needed to make the message more precise:

- End with a period
- Consider singular and plural

```javascript
const message = sprintf(
	// translators: %d: the name of the block that has been moved
	_n(
		'%d block moved.',
		'%d blocks moved.',
		clientIds.length
	),
	clientIds.length
);
speak( message );
```

The code is similar to [here](https://github.com/WordPress/gutenberg/blob/dc4bd0502d9ef354a3f9556d1a36aa1a60d526b8/packages/block-editor/src/components/inserter/hooks/use-insertion-point.js#L183-L188).

**Comment by n2erjo00** [comment]
@t-hamano Added period and logic for singular/plural

File: packages/block-editor/src/components/block-tools/index.js [hunk 1]
```diff
@@ -92,19 +92,31 @@ export default function BlockTools( {
...
+					moveBlocksDown( clientIds, rootClientId );
+				}
+				const blockLength = Array.isArray( clientIds )
+					? clientIds.length
+					: 1;
+				const message = sprintf(
+					// translators: %s: term "block" in either singular or plural form.
+					_n( '%s moved.', '%s moved.', blockLength ),
+					blockLength > 1 ? __( 'Blocks' ) : __( 'Block' )
+				);
```
Comment (t-hamano) [inline packages/block-editor/src/components/block-tools/index.js]: Sorry for the insufficient explanation. In this implementation, we cannot know how many blocks have actually been moved.

What I wanted to say is to tell how many blocks have been moved. That is, the code I proposed in [this comment](https://github.com/WordPress/gutenberg/pull/64966#issuecomment-2443058839).

I think this code is consistent with the implementation below:

https://github.com/WordPress/gutenberg/blob/986fc5519b9455c2683f519abe4d6446111d5e38/packages/block-editor/src/components/inserter/hooks/use-insertion-point.js#L183-L187

https://github.com/WordPress/gutenberg/blob/986fc5519b9455c2683f519abe4d6446111d5e38/packages/editor/src/components/block-manager/index.js#L87-L95

Comment (n2erjo00) [reply packages/block-editor/src/components/block-tools/index.js]: @t-hamano Added more consistent messaging

**Review by t-hamano — APPROVED** [review]
LGTM! It seems to work as expected.

The video below shows how NVDA reads it out loud:

https://github.com/user-attachments/assets/82662e53-1289-4457-992d-29164256348a

**Comment by t-hamano** [comment]
Note: I would like to close #24894, but add the people who worked on it as props.

```
Co-authored-by: enricobattocchi <lopo@git.wordpress.org>
Co-authored-by: youknowriad <youknowriad@git.wordpress.org>
Co-authored-by: mtias <matveb@git.wordpress.org>
```

**Comment by elettrona** [comment]
I have to report an issue - at least in Italian version.
When moving blocks with keystrokes - ctrl+alt+t up, ctrl+alt+y down, it just says "one block moved". Nothing else.
I'd expect my screen reader to say "block 1 moved down to row 2"
and if I move it further it would say "block 2 moved down to row 3"... same for moving upwards.
"block moved" as announcement is meaningless.

**Comment by elettrona** [comment]
I have reported it on #61168 issue some months ago and I'm pleased it has been worked on, but I hope the notification will be made more detailed.

**Comment by afercia** [comment]
HI @talksina thanks for your feedback. as @t-hamano mentioned on the related issue https://github.com/WordPress/gutenberg/issues/61168#issuecomment-2493705370, in a previous comment on this pull requests https://github.com/WordPress/gutenberg/pull/64966#issuecomment-2406880252 whether to use a shorter message or a more informative one was discussed. The shorter message option prevailed. However, maybe it's worth reconsidering. One of the assumption was:

When using the buttons in the block toolbar, the button provides a description with the position information. For example, the aria label and description would be:
- label: Move down
- description: Move Paragraph block from position 1 down to position 2 

As such, in this specific case, the position information is provided _before_ moving the block. The shorter confirmation message would make sense in thei case. However, I think a message isn't provided at all when using the buttons. 

Instead, when using the keyboard shortcuts, there's no information about the starting position. There's only the short confirmation message

**Comment by afercia** [comment]
One more thing to fix I just noticed: when a block can't be moved because:
- it's at the first position and can't be moved up
- it's at the last position and can't be moved down
- it's the only block

The keyboard shortcuts are still triggered and the speak message still announces "Block moved" even though it didn't.
I will create a new issue to summarize all the remaining points to address.

**Comment by afercia** [comment]
New issue created: https://github.com/WordPress/gutenberg/issues/67237

**Pull request merged**